src/ILCompiler.Compiler/src/Compiler/TypeInitialization.cs

```diff
@@ -42,4 +48,99 @@ private static bool HasEagerConstructorAttribute(TypeDesc type)
                 || mdType.HasCustomAttribute("System.Runtime.CompilerServices", "EagerStaticClassConstructionAttribute"));
         }
     }
+
+    public class EagerConstructorComparer : IComparer<DependencyAnalysis.IMethodNode>
+    {
+        private bool TryGetConstructionOrder(MetadataType type, out int order)
+        {
+            // Order is defined by an integer in EagerOrderedStaticConstructorAttribute.
+            // If this is a type with EagerStaticClassConstructionAttribute, order is not
+            // defined explicitly.
+
+            type = (MetadataType)type.GetTypeDefinition();
+
+            EcmaType ecmaType = (EcmaType)type;
+            MetadataReader metadataReader = ecmaType.MetadataReader;
+
+            foreach (var attributeHandle in metadataReader.GetTypeDefinition(ecmaType.Handle).GetCustomAttributes())
+            {
+                EntityHandle attributeType, attributeCtor;
+                if (!metadataReader.GetAttributeTypeAndConstructor(attributeHandle,
+                    out attributeType, out attributeCtor))
+                {
+                    continue;
+                }
+
+                StringHandle namespaceHandle, nameHandle;
+                if (!metadataReader.GetAttributeTypeNamespaceAndName(attributeType,
+                   out namespaceHandle, out nameHandle))
+                {
+                    continue;
+                }
+
+                if (metadataReader.StringComparer.Equals(namespaceHandle, "System.Runtime.CompilerServices")
+                    && metadataReader.StringComparer.Equals(nameHandle, "EagerOrderedStaticConstructorAttribute"))
+                {
+                    var constructor = ecmaType.EcmaModule.GetMethod(attributeCtor);
+                    if (constructor.Signature.Length != 1)
+                        throw new BadImageFormatException();
+
+                    var attributeBlob = metadataReader.GetBlobReader(metadataReader.GetCustomAttribute(attributeHandle).Value);
+                    if (attributeBlob.ReadInt16() != 1)
+                        throw new BadImageFormatException();
+
+                    order = attributeBlob.ReadInt32();
+                    return true;
+                }
+            }
+
+            Debug.Assert(type.HasCustomAttribute("System.Runtime.CompilerServices", "EagerStaticClassConstructionAttribute"));
+            order = 0;
+            return false;
+        }
+
+        public int Compare(DependencyAnalysis.IMethodNode x, DependencyAnalysis.IMethodNode y)
+        {
+            var typeX = (MetadataType)x.Method.OwningType;
+            var typeY = (MetadataType)y.Method.OwningType;
+
+            int orderX;
+            bool hasOrderX = TryGetConstructionOrder(typeX, out orderX);
```

Comment:
This may be simpler if we assume that the order for unordered ones is `-1`:
```
if (orderX != orderY)
{
    result = Comparer<int>.Default.Compare(orderX, orderY);
}
else
{
    result = String.Compare(...);
}
```